vms/platformvm/state/stakers_model_storage_test.go

```diff
@@ -0,0 +1,966 @@
+	_ commands.Command = (*deleteCurrentValidatorCommand)(nil)
+	_ commands.Command = (*putCurrentDelegatorCommand)(nil)
+	_ commands.Command = (*deleteCurrentDelegatorCommand)(nil)
+	_ commands.Command = (*addTopDiffCommand)(nil)
+	_ commands.Command = (*applyAndCommitBottomDiffCommand)(nil)
+	_ commands.Command = (*rebuildStateCommand)(nil)
+
+	commandsCtx = buildStateCtx()
+)
+
+// TestStateAndDiffComparisonToStorageModel verifies that a production-like
+// system made of a stack of Diffs built on top of a State conforms to
+// our stakersStorageModel. It achieves this by:
+//  1. randomly generating a sequence of stakers writes as well as
+//     some persistence operations (commit/diff apply),
+//  2. applying the sequence to both our stakersStorageModel and the production-like system.
+//  3. checking that both stakersStorageModel and the production-like system have
+//     the same state after each operation.
+//
+// The following invariants are required for stakers state to properly work:
+//  1. No stakers add/update/delete ops are performed directly on baseState, but on at least a diff
+//  2. Any number of stakers ops can be carried out on a single diff
+//  3. Diffs work in FIFO fashion: they are added on top of current state and only
+//     bottom diff is applied to base state.
+//  4. The bottom diff applied to base state is immediately committed.
+func TestStateAndDiffComparisonToStorageModel(t *testing.T) {
+	properties := gopter.NewProperties(nil)
+
+	// // to reproduce a given scenario do something like this:
+	// parameters := gopter.DefaultTestParametersWithSeed(1688641048828490074)
+	// properties := gopter.NewProperties(parameters)
+
+	properties.Property("state comparison to storage model", commands.Prop(stakersCommands))
+	properties.TestingRun(t)
+}
+
+// stakersCommands creates/destroy the system under test and generates
+// commands and initial states (stakersStorageModel)
+var stakersCommands = &commands.ProtoCommands{
+	NewSystemUnderTestFunc: func(initialState commands.State) commands.SystemUnderTest {
+		model := initialState.(*stakersStorageModel)
+
+		baseDBManager := manager.NewMemDB(version.Semantic1_0_0)
+		baseDB := versiondb.New(baseDBManager.Current().Database)
+		baseState, err := buildChainState(baseDB, nil)
+		if err != nil {
+			panic(err)
+		}
+
+		// fillup baseState with model initial content
+		for _, staker := range model.currentValidators {
+			baseState.PutCurrentValidator(staker)
+		}
+		for _, delegators := range model.currentDelegators {
+			for _, staker := range delegators {
+				baseState.PutCurrentDelegator(staker)
+			}
+		}
+		for _, staker := range model.pendingValidators {
+			baseState.PutPendingValidator(staker)
+		}
+		for _, delegators := range model.currentDelegators {
+			for _, staker := range delegators {
+				baseState.PutPendingDelegator(staker)
+			}
+		}
+		if err := baseState.Commit(); err != nil {
+			panic(err)
+		}
+
+		return newSysUnderTest(baseDB, baseState)
```

Comment:
The system-under-test constructor. We make sure it is filled with model initial content (even if model is currently empty at start)